client/X11/xf_cliprdr.c

```diff
@@ -1663,6 +1928,334 @@ static UINT xf_cliprdr_clipboard_file_range_failure(wClipboardDelegate* delegate
+	ArrayList_Unlock(clipboard->ino_list);
+	if (!node)
+	{
+		fuse_reply_err(req, ENONET);
+		return;
+	}
+	else if ((node->st_mode & S_IFDIR) != 0)
+	{
+		fuse_reply_err(req, EISDIR);
+		return;
+	}
+
+	xfCliprdrFuseStream* stream = (xfCliprdrFuseStream*)calloc(1, sizeof(xfCliprdrFuseStream));
+
+	ArrayList_Lock(clipboard->stream_list);
+	stream->req = req;
+	stream->stream_id = clipboard->current_stream_id;
+	clipboard->current_stream_id++;
+	ArrayList_Add(clipboard->stream_list, stream);
+	ArrayList_Unlock(clipboard->stream_list);
+
+	UINT32 nPositionLow = (off >> 0) & 0xFFFFFFFF;
+	UINT32 nPositionHigh = (off >> 32) & 0xFFFFFFFF;
+
+	xf_cliprdr_send_client_file_contents(clipboard, stream->stream_id, node->lindex, nPositionLow,
+	                                     nPositionHigh, size);
+}
+
+static void xf_cliprdr_fuse_lookup(fuse_req_t req, fuse_ino_t parent, const char* name)
+{
+	size_t index;
+	size_t count;
+	size_t child_ino;
+	BOOL found = FALSE;
+	struct fuse_entry_param e;
+	xfCliprdrFuseInode* parent_node;
+	xfCliprdrFuseInode* child_node;
+	xfClipboard* clipboard = (xfClipboard*)fuse_req_userdata(req);
+
+	ArrayList_Lock(clipboard->ino_list);
+	parent_node = (xfCliprdrFuseInode*)ArrayList_GetItem(clipboard->ino_list, parent - 1);
+	ArrayList_Unlock(clipboard->ino_list);
+	if (!parent_node || !parent_node->child_inos)
+	{
+		fuse_reply_err(req, ENOENT);
+		return;
+	}
+
+	ArrayList_Lock(parent_node->child_inos);
+	count = ArrayList_Count(parent_node->child_inos);
+	for (index = 0; index < count; index++)
+	{
+		child_ino = (size_t)ArrayList_GetItem(parent_node->child_inos, index);
+		ArrayList_Lock(clipboard->ino_list);
+		child_node = (xfCliprdrFuseInode*)ArrayList_GetItem(clipboard->ino_list, child_ino - 1);
+		ArrayList_Unlock(clipboard->ino_list);
+		if (child_node && strcmp(name, child_node->name) == 0)
+		{
+			found = TRUE;
+			break;
+		}
+	}
+	ArrayList_Unlock(parent_node->child_inos);
+	if (found == TRUE)
+	{
+		memset(&e, 0, sizeof(e));
+		e.ino = child_node->ino;
+		e.attr_timeout = 1.0;
+		e.entry_timeout = 1.0;
+		e.attr.st_ino = child_node->ino;
+		e.attr.st_mode = child_node->st_mode;
+		e.attr.st_nlink = 1;
+		e.attr.st_size = child_node->st_size;
+		e.attr.st_mtime = child_node->st_mtim.tv_sec;
+		fuse_reply_entry(req, &e);
+	}
+	else
+	{
+		fuse_reply_err(req, ENOENT);
+	}
+}
+
+static void xf_cliprdr_fuse_opendir(fuse_req_t req, fuse_ino_t ino, struct fuse_file_info* fi)
+{
+	xfCliprdrFuseInode* node;
+	xfClipboard* clipboard = (xfClipboard*)fuse_req_userdata(req);
+
+	ArrayList_Lock(clipboard->ino_list);
+	node = (xfCliprdrFuseInode*)ArrayList_GetItem(clipboard->ino_list, ino - 1);
+	ArrayList_Unlock(clipboard->ino_list);
+	if (!node)
+	{
+		fuse_reply_err(req, ENOENT);
+	}
+	else if ((node->st_mode & S_IFDIR) == 0)
+	{
+		fuse_reply_err(req, ENOTDIR);
+	}
+	else
+	{
+		fuse_reply_open(req, fi);
+	}
+}
+
+static struct fuse_lowlevel_ops xf_cliprdr_fuse_oper = {
+	.lookup = xf_cliprdr_fuse_lookup,
```

Comment:
You don't need an implementation for `lookup` and `opendir` (not necessary, can be removed).

Reply:
i think no. otherwise fuse do not know relationship between  ino number with filename.

Reply:
Ah, I just see that the function signatures for `fuse_operations` and `fuse_lowlevel_ops` are not the same (`fuse_operations` works on the filepath and not on the inode while `fuse_lowlevel_ops` uses the inode), so nvm.